### DISCUSSION TITLE
Inmutabilidad, vale la pena?

### DISCUSSION THREAD

**Discussion by Guernik** [dds-jv/foro#121] [Consultas de objetos] [answered]
Hola! Basicamente esa es la consulta.
[Dadas las abrumadoras bondades](http://www.javapractices.com/topic/TopicAction.do?Id=29) de la inmutabilidad, me preguntaba si vale la pena hacer el diseño de esa manera. 
Soy conciente de que no es tan sencillo! Sobretodo con las collections :fearful:.

Saludos :smiley: !

**Comment by roli93** [comment, marked as answer]
Hola!

Como todo en esta ~materia~ vida, _depende_

Hablar de hacer "el diseño" inmutable, como si de un todo completo se hablara, es algo peligroso y confuso. Es como decir "Voy a hacer mi diseño simple". Sí, tal vez esté bueno que busques privilegiar esa cualidad, pero recordemos que en DDS nunca hablamos de absolutos. Siempre de "más o menos que". Con la inmutabilidad, si bien no es una cualidad de diseño per se, es cierto que suele traer aparejadas otras que sí lo son, y si bien no es cualitativamente regulable (no sé si tendría sentido hablar de que algo sea 'medio inmutabe') sí lo es en forma cuantitativa (puedo hacer muchos objetos inmutables, o pocos, o algunos). Y eso es lo que deberías regular.

Muy probablemente, si intentás forzar 100% de inmutabilidad en un lenguaje como Java que tiene muchas nociones de objetos e imperativas, que descansan fuerte sobre el efecto de lado, vayas a comprarte un problema y sufras mucho al pedo, porque en muchos casos los beneficios van a ser nulos. Sí es una buena idea tal vez apenas modeles una clase, tratar de arrancar pensándola como inmutable, no ponerle setters y recién cuando veas la necesidad de mutarla preguntarte: "Qué alterntivas tengo?" "Qué pierdo?" "Qué gano"?. 

Probablemente si estás pensando en mutar un atributo para almacenar valores intermedios de un procesamiento interno, sea una pésima idea, porque harías tu código más dependiente el orden, mas interacoplado y menos claro de seguir (y ni hablemos de si llega a haber problemas de concurrencia :P). Ahora si el dominio dice "El usuario debe poder editar el medio de pago de la venta". Muy probablemente sea infinito más fácil mutar la referencia que crear una nueva venta que apunte a otro medio de pago. 

Respecto de las colecciones per se, si usas la API de streams de Java (`miLista.stream().map(...)`, etc.) la mayoría de los métodos devuelven una lista nueva. Respecto de los elementos internos de la lista, es el problema que relaté en el párrafo anterior (si creo una nueva venta tengo que sacar la venta anterior y agregar la nueva y encima poder identificar que 'son la misma') , y probablemente no vale la pena. La 'identidad propia' de los objetos es una ventaja muy grande que nos da el paradigma y no queremos perderla si no es por una muy buena razón.

Por último, respecto del tema, si llegan a ver algo de `java.Cloneable` ***eviten bajo todo concepto usarlo***. No es lo que parece, no es un copy mágico que clona objetos como uno quisiera y todos los problemas de hacerlo a mano siguen mayormente ahí.

Espero haber sido de ayuda (aunque básicamente dije "Ni idea, depende, fijate", pero bueno, ese es el tema con esta materia xD)

Saludos!

**Comment by gastonprieto** [comment]
@Guernik te sirvio ?

**Comment by Guernik** [comment]
Hola, sí, perdón que lo dejé colgado. Sí me sirvió. Basicamente la respuesta es "depende"  :rofl:
Gracias!

**Comment by gastonprieto** [comment]
Exacto ;)

Cierro el issue entonces.
